### PR TITLE
Quit: delay process termination until cleanup actions have run (#3270)

### PR DESCRIPTION
**Pull request by timbertson**
Fixes https://github.com/onivim/oni2/issues/3270.

Needs to be applied alongside https://github.com/revery-ui/revery/pull/1053 due to breaking API change in revery.

### CODE REVIEW

**Comment by timbertson** [comment]
I'm not really sure what to do with dependencies - should I pin `revery` to my fork temporarily so CI will pass?

I edited package.json manually to do:

```diff
diff --git package.json package.json
index ebe544c1..c5b5687b 100644
--- package.json
+++ package.json
@@ -253,7 +253,7 @@
     "revery-terminal": "*"
   },
   "resolutions": {
-    "revery": "revery-ui/revery#26e8b73",
+    "revery": "timbertson/revery#4433506",
     "esy-skia": "revery-ui/esy-skia#91c98f6",
     "isolinear": "revery-ui/isolinear#53fc4eb",
     "revery-terminal": "revery-ui/revery-terminal#a9cb168",
```

But then `esy solve` produced a terrifying number of changes, so I didn't want to commit that:

```
 14 files changed, 104 insertions(+), 935 deletions(-)
```

**Comment by bryphe** [comment]
Awesome, thanks @timbertson !

> I'm not really sure what to do with dependencies - should I pin revery to my fork temporarily so CI will pass?

I created a PR to pick up the revery dependency (https://github.com/onivim/oni2/pull/3334) - once that's in, this should be unblocked after merging. Just helpful to have the incremental change, and then it will keep this PR scoped to the new quit behavior.

File: src/bin_editor/Oni2_editor.re [hunk 1]
```diff
@@ -507,11 +507,14 @@ switch (eff) {
...
+    Revery.App.onBeforeQuit(app, _code => {
       if (!currentState^.isQuitting) {
-        dispatch(Model.Actions.Quit(true));
-      }
-    )
+        dispatch(Model.Actions.ReallyQuitting);
+      };
+      // We perform asynchronous cleanup, and
+      // exit(0) ourselves when that's complete
+      Revery.App.PreventQuit
```
Comment (bryphe) [inline src/bin_editor/Oni2_editor.re:516]: Thanks for working through this! The lifecycle-management pieces are tricky - I really like the way you investigated it by evaluating the different possibilities in #3270  and then implementing the changes in Revery to support this.

File: src/Store/ExtensionClientStoreConnector.re [hunk 2]
```diff
@@ -25,7 +25,10 @@ let start = (extensions, extHostClient: Exthost.Client.t) => {
       ~name="exthost.registerQuitCleanup", dispatch =>
       dispatch(
         Actions.RegisterQuitCleanup(
-          () => Exthost.Client.terminate(extHostClient),
+          () => {
+            Log.debug("terminating exthost");
+            Exthost.Client.terminate(extHostClient)
```
Comment (bryphe) [inline src/Store/ExtensionClientStoreConnector.re:30]: Excited for these warnings to go away 🎉

**Comment by timbertson** [comment]
I see you've added some merge commits etc, thanks! I'm not entirely sure how to navigate CI so I'm happy to leave it to you, but please let me know if there's something I can help with.

**Comment by bryphe** [comment]
Hi @timbertson ! Sorry for the slow reply. Changes look good; CI issue doesn't seem to be related to these changes.

Thank you for the contribution; I tagged this as meeting the criteria for our [License Key Bounty](https://onivim.github.io/docs/for-developers/free-license). Had to dive across the stack and implement a fix in Revery for this 😄 Is the e-mail on your GitHub profile a good one to send it to?

**Comment by timbertson** [comment]
Thanks @bryphe , you've made it a pleasure to contribute, and I hope I'll be back soon with more 😀 

I (happily) paid for a license a while back, but if you want to shoot one to my profile email address maybe I'll share the love with a friend :)

**Comment by bryphe** [comment]
Hi @timbertson - awesome, really appreciate it :smile: 

Just e-mailed you a coupon code you can share with your friend -

**Pull request merged**